File: source/nds.cpp
#include "m3_env.h"	
#include <nds.h>
#include <nf_lib.h>

m3ApiRawFunction(m3_swiWaitForVBlank) {
    swiWaitForVBlank();
    m3ApiSuccess();
}

m3ApiRawFunction(_print)
{
    m3ApiGetArgMem (const uint8_t *, buf)
    iprintf((char *)buf);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_Set2D) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint8_t, model)
    NF_Set2D(screen, model);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_SetRootFolder) {
    m3ApiGetArgMem (const uint8_t *, root)
    NF_SetRootFolder((char*) root);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_InitTiledBgBuffers) {
    NF_InitTiledBgBuffers();
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_InitTiledBgSys) {
    m3ApiGetArg (uint8_t, screen)
    NF_InitTiledBgSys(screen);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_LoadTiledBg) {
    m3ApiGetArgMem (const uint8_t *, path)
    m3ApiGetArgMem (const uint8_t *, name)
    m3ApiGetArg (uint32_t, width)
    m3ApiGetArg (uint32_t, height)
    NF_LoadTiledBg((char*)path, (char*)name, width, height);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_CreateTiledBg) {
    m3ApiGetArg (uint32_t, a)
    m3ApiGetArg (uint32_t, b)
    m3ApiGetArgMem (const uint8_t *, c)
    NF_CreateTiledBg(a,b,(char*)c);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_InitSpriteBuffers) {
    NF_InitSpriteBuffers();
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_InitSpriteSys) {
    m3ApiGetArg (uint8_t, screen)
    NF_InitSpriteSys(screen);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_InitTextSys) {
    m3ApiGetArg (uint8_t, screen)
    NF_InitTextSys(screen);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_LoadTextFont) {
    m3ApiGetArgMem (const uint8_t *, file)
    m3ApiGetArgMem (const uint8_t *, name)
    m3ApiGetArg (uint32_t, width)
    m3ApiGetArg (uint32_t, height)
    m3ApiGetArg (uint32_t, rotation)
    NF_LoadTextFont((char*)file, (char*)name, width, height, rotation);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_CreateTextLayer) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint8_t, layer)
    m3ApiGetArg (uint8_t, rotation)
    m3ApiGetArgMem (const uint8_t *, name)
    NF_CreateTextLayer(screen, layer, rotation, (char*)name);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_WriteText) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint8_t, layer)
    m3ApiGetArg (uint8_t, x)
    m3ApiGetArg (uint8_t, y)
    m3ApiGetArgMem (const uint8_t *, text)
    NF_WriteText(screen, layer, x, y, (char*)text);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_UpdateTextLayers) {
    NF_UpdateTextLayers();
    m3ApiSuccess();
}

m3ApiRawFunction(m3_consoleDemoInit) {
    consoleDemoInit();
    m3ApiSuccess();
};

m3ApiRawFunction(m3_consoleClear) {
    consoleClear();
    m3ApiSuccess();
};

m3ApiRawFunction(m3_NF_LoadSpriteGfx) {
    m3ApiGetArgMem (const uint8_t *, a)
    m3ApiGetArg (uint8_t, b)
    m3ApiGetArg (uint8_t, c)
    m3ApiGetArg (uint8_t, d)
    NF_LoadSpriteGfx((char*)a, b, c, d); 
    m3ApiSuccess();
};

m3ApiRawFunction(m3_NF_LoadSpritePal) {
    m3ApiGetArgMem (const uint8_t *, a)
    m3ApiGetArg (uint8_t, b)
    NF_LoadSpritePal((char*)a, b); 
    m3ApiSuccess();
};

m3ApiRawFunction(m3_NF_VramSpriteGfx) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint16_t, ram)
    m3ApiGetArg (uint16_t, vram)
    m3ApiGetArg (bool, keepframes)
    NF_VramSpriteGfx(screen, ram, vram, keepframes); 
    m3ApiSuccess();
};

m3ApiRawFunction(m3_NF_VramSpritePal) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint8_t, id)
    m3ApiGetArg (uint8_t, slot)
    NF_VramSpritePal(screen, id, slot); 
    m3ApiSuccess();
};


m3ApiRawFunction(m3_NF_CreateSprite) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint8_t, id)
    m3ApiGetArg (uint8_t, gfx)
    m3ApiGetArg (uint8_t, pal)
    m3ApiGetArg (uint8_t, x)
    m3ApiGetArg (uint8_t, y)
    NF_CreateSprite(screen, id, gfx, pal, x, y); 
    m3ApiSuccess();
};

m3ApiRawFunction(m3_NF_SpriteOamSet) {
    m3ApiGetArg (uint8_t, screen)
    NF_SpriteOamSet(screen); 
    m3ApiSuccess();
};

m3ApiRawFunction(m3_oamUpdate) {
    oamUpdate(&oamMain);
    m3ApiSuccess();
}

m3ApiRawFunction(m3_NF_ClearTextLayer) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint8_t, layer)
    NF_ClearTextLayer(screen, layer); 
    m3ApiSuccess();
};

m3ApiRawFunction(m3_NF_DeleteSprite) {
    m3ApiGetArg (uint8_t, screen)
    m3ApiGetArg (uint8_t, layer)
    NF_DeleteSprite(screen, layer); 
    m3ApiSuccess();
};

m3ApiRawFunction(m3_keysHeld) {
    m3ApiReturnType (uint32_t)
    m3ApiReturn(keysHeld());
};

m3ApiRawFunction(m3_rand) {
    m3ApiReturnType (uint32_t)
    m3ApiReturn(rand());
};

m3ApiRawFunction(m3_scanKeys) {
    scanKeys();
    m3ApiSuccess();
};

void LinkNDSFunctions(IM3Module module) {
    m3_LinkRawFunction (module, "nds", "swiWaitForVBlank", "v()", &m3_swiWaitForVBlank);
    m3_LinkRawFunction (module, "nds", "_print", "v(i)", &_print);
    m3_LinkRawFunction (module, "nds", "NF_Set2D", "v(ii)", &m3_NF_Set2D);
    m3_LinkRawFunction (module, "nds", "_NF_SetRootFolder", "v(i)", &m3_NF_SetRootFolder);
    m3_LinkRawFunction (module, "nds", "NF_InitTiledBgBuffers", "v()", &m3_NF_InitTiledBgBuffers);
    m3_LinkRawFunction (module, "nds", "NF_InitTiledBgSys", "v(i)", &m3_NF_InitTiledBgSys);
    m3_LinkRawFunction (module, "nds", "_NF_LoadTiledBg", "v(iiii)", &m3_NF_LoadTiledBg);
    m3_LinkRawFunction (module, "nds", "_NF_CreateTiledBg", "v(iii)", &m3_NF_CreateTiledBg);
    m3_LinkRawFunction (module, "nds", "NF_InitSpriteBuffers", "v()", &m3_NF_InitSpriteBuffers);
    m3_LinkRawFunction (module, "nds", "NF_InitSpriteSys", "v(i)", &m3_NF_InitSpriteSys);
    m3_LinkRawFunction (module, "nds", "NF_InitTextSys", "v(i)", &m3_NF_InitTextSys);
    m3_LinkRawFunction (module, "nds", "_NF_LoadTextFont", "v(iiiii)", &m3_NF_LoadTextFont);
    m3_LinkRawFunction (module, "nds", "_NF_CreateTextLayer", "v(iiii)", &m3_NF_CreateTextLayer);
    m3_LinkRawFunction (module, "nds", "_NF_WriteText", "v(iiiii)", &m3_NF_WriteText);
    m3_LinkRawFunction (module, "nds", "NF_UpdateTextLayers", "v()", &m3_NF_UpdateTextLayers);
    m3_LinkRawFunction (module, "nds", "consoleDemoInit", "v()", &m3_consoleDemoInit);
    m3_LinkRawFunction (module, "nds", "consoleClear", "v()", &m3_consoleClear);
    m3_LinkRawFunction (module, "nds", "_NF_LoadSpriteGfx", "v(iiii)", &m3_NF_LoadSpriteGfx);
    m3_LinkRawFunction (module, "nds", "_NF_LoadSpritePal", "v(ii)", &m3_NF_LoadSpritePal);
    m3_LinkRawFunction (module, "nds", "NF_VramSpriteGfx", "v(iiii)", &m3_NF_VramSpriteGfx);
    m3_LinkRawFunction (module, "nds", "NF_VramSpritePal", "v(iii)", &m3_NF_VramSpritePal);
    m3_LinkRawFunction (module, "nds", "NF_CreateSprite", "v(iiiiii)", &m3_NF_CreateSprite);
    m3_LinkRawFunction (module, "nds", "NF_SpriteOamSet", "v(i)", &m3_NF_SpriteOamSet);
    m3_LinkRawFunction (module, "nds", "oamUpdate", "v()", &m3_oamUpdate);
    m3_LinkRawFunction (module, "nds", "NF_ClearTextLayer", "v(ii)", &m3_NF_ClearTextLayer);
    m3_LinkRawFunction (module, "nds", "NF_DeleteSprite", "v(ii)", &m3_NF_DeleteSprite);
    m3_LinkRawFunction (module, "nds", "keysHeld", "i()", &m3_keysHeld);
    m3_LinkRawFunction (module, "nds", "scanKeys", "v()", &m3_scanKeys);
    m3_LinkRawFunction (module, "nds", "rand", "i()", &m3_rand);
}

File: assemblyscript/assembly/nds.ts
export declare function swiWaitForVBlank(): void;

const encode = (str: string): ArrayBuffer => String.UTF8.encode(str + "\0");

declare function _print(ptr: ArrayBuffer): void;
export function print(str: string): void {
  _print(encode(str));
}

export declare function NF_Set2D(screen: u8, model: u8): void;

export declare function _NF_SetRootFolder(a: ArrayBuffer): void;
export function NF_SetRootFolder(root: string): void {
  _NF_SetRootFolder(encode(root));
}

export declare function NF_InitTiledBgBuffers(): void;
export declare function NF_InitTiledBgSys(screen: u8): void;

declare function _NF_LoadTiledBg(
  file: ArrayBuffer,
  name: ArrayBuffer,
  width: u32,
  height: u32
): void;
export function NF_LoadTiledBg(
  file: string,
  name: string,
  width: u32,
  height: u32
): void {
  _NF_LoadTiledBg(encode(file), encode(name), width, height);
}

export declare function _NF_CreateTiledBg(
  screen: u32,
  layer: u32,
  name: ArrayBuffer
): void;
export function NF_CreateTiledBg(screen: u32, layer: u32, name: string): void {
  _NF_CreateTiledBg(screen, layer, encode(name));
}

export declare function NF_InitSpriteBuffers(): void;
export declare function NF_InitSpriteSys(screen: u8): void;
export declare function NF_InitTextSys(screen: u8): void;

declare function _NF_LoadTextFont(
  file: ArrayBuffer,
  name: ArrayBuffer,
  width: u32,
  height: u32,
  rotation: u32
): void;
export function NF_LoadTextFont(
  file: string,
  name: string,
  width: u32,
  height: u32,
  rotation: u32
): void {
  _NF_LoadTextFont(encode(file), encode(name), width, height, rotation);
}

declare function _NF_CreateTextLayer(
  screen: u8,
  layer: u8,
  rotation: u8,
  name: ArrayBuffer
): void;
export function NF_CreateTextLayer(
  screen: u8,
  layer: u8,
  rotation: u8,
  name: string
): void {
  _NF_CreateTextLayer(screen, layer, rotation, encode(name));
}

declare function _NF_WriteText(
  screen: u8,
  layer: u8,
  x: u16,
  y: u16,
  text: ArrayBuffer
): void;
export function NF_WriteText(
  screen: u8,
  layer: u8,
  x: u16,
  y: u16,
  text: string
): void {
  _NF_WriteText(screen, layer, x, y, encode(text));
}

export declare function NF_UpdateTextLayers(): void;

export declare function consoleDemoInit(): void;
export declare function consoleClear(): void;

declare function _NF_LoadSpriteGfx(
  file: ArrayBuffer,
  id: u16,
  width: u16,
  height: u16
): void;
export function NF_LoadSpriteGfx(
  file: string,
  id: u16,
  width: u16,
  height: u16
): void {
  _NF_LoadSpriteGfx(encode(file), id, width, height);
}

declare function _NF_LoadSpritePal(file: ArrayBuffer, id: u8): void;
export function NF_LoadSpritePal(file: string, id: u8): void {
  _NF_LoadSpritePal(encode(file), id);
}

export declare function NF_VramSpriteGfx(
  screen: u8,
  vram: u16,
  ram: u16,
  keepframes: boolean
): void;
export declare function NF_VramSpritePal(screen: u8, id: u8, slot: u8): void;

export declare function NF_CreateSprite(
  screen: u8,
  id: u8,
  gfx: u16,
  pal: u8,
  x: i16,
  y: i16
): void;
export declare function NF_SpriteOamSet(screen: u8): void;
export declare function oamUpdate(): void;
export declare function NF_ClearTextLayer(screen: u8, layer: u8): void;
export declare function NF_DeleteSprite(screen: u8, layer: u8): void;
export declare function keysHeld(): u32;
export declare function scanKeys(): void;

export enum KEYS {
  SELECT = 1 << 2,
  START = 1 << 3,
  RIGHT = 1 << 4,
  LEFT = 1 << 5,
  UP = 1 << 6,
  DOWN = 1 << 7,
}

export declare function rand(): u8;
